tests/tests.py

```diff
@@ -1,199 +1,594 @@
 #!/usr/bin/env python3
 
 import argparse
-import configparser
+import contextlib
+import enum
 import hashlib
+import itertools
 import os
 import sys
 import tempfile
+import unittest.mock
 import zipfile
 
 sys.path.append(os.path.join(sys.path[0], '..'))
-from avbroot import main
-from avbroot import ota
-from avbroot import util
+from avbroot import external, ota, util
+from avbroot.main import PARTITION_PRIORITIES, PATH_PAYLOAD
+import avbroot.main
+import ota_utils
+import update_metadata_pb2
 
+import config
 import downloader
 
+FILES_DIR = os.path.join(sys.path[0], 'files')
+TEST_KEY_PREFIX = os.path.join(sys.path[0], 'keys', 'TEST_KEY_DO_NOT_USE_')
 
-def url_filename(url):
-    return url.split('/')[-1]
+PARTITIONS_TO_PRESERVE = set(sum(PARTITION_PRIORITIES.values(), ()))
 
 
-def device_configs(args, config, filter=True):
-    for section in config.sections():
-        assert section == os.path.basename(section)
+def hash_zeroes(hasher, size):
+    null_buf = memoryview(b'\0' * 65536)
 
-        if section.startswith('device.'):
-            name = section.removeprefix('device.')
-            if filter and args.device and name not in args.device:
-                continue
+    while size > 0:
+        n = min(size, len(null_buf))
+        hasher.update(null_buf[:n])
+        size -= n
 
-            yield name, config[section]
 
+def hash_file_path(path):
+    print('Calculating hash of', path)
 
-def verify_checksum(path, sha256_hex):
     with open(path, 'rb') as f:
-        hasher = util.hash_file(f, hashlib.sha256())
+        return util.hash_file(f, hashlib.sha256())
+
+
+def verify_hash(path, sha256_hex):
+    hasher = hash_file_path(path)
 
     if hasher.digest() != bytes.fromhex(sha256_hex):
         raise Exception(f'{path}: expected sha256 {sha256_hex}, '
                         f'but have {hasher.hexdigest()}')
 
 
-def download(directory, url, sha256_hex, revalidate=False):
-    path = os.path.join(directory, url_filename(url))
-    validate = True
+def merge_overlapping(sections):
+    '''
+    Sort and merge overlapping intervals.
+    '''
+
+    result = []
+
+    for start, end in sorted(sections):
+        if result and start <= result[-1][1]:
+            new_section = (result[-1][0], end)
+            result.pop()
+            result.append(new_section)
+        else:
+            result.append((start, end))
+
+    return result
+
+
+def exclusion_to_inclusion(holes, start, end):
```

Comment:
Although not applicable for our use case; if we want this to be a universal exclusion to inclusion function, the cases where `start == exclusions[0]` and `end == exclusions[-1]`should not be in the output.

Reply:
Good catch, I'll fix that.